src/bitpacker4x.rs

```diff
@@ -0,0 +1,254 @@
+    use std::arch::x86_64::_mm_slli_epi32 as left_shift_32;
+    use std::arch::x86_64::_mm_or_si128 as op_or;
+    use std::arch::x86_64::_mm_and_si128 as op_and;
+    use std::arch::x86_64::_mm_lddqu_si128 as load_unaligned;
+    use std::arch::x86_64::_mm_storeu_si128 as store_unaligned;
+    use std::arch::x86_64::{_mm_add_epi32,  _mm_shuffle_epi32, _mm_srli_si128, _mm_sub_epi32, _mm_slli_si128, _mm_cvtsi128_si32};
+
+    #[allow(non_snake_case)]
+    unsafe fn or_collapse_to_u32(accumulator: DataType) -> u32 {
+        let a__b__c__d_ = accumulator;
+        let ______a__b_ = _mm_srli_si128(a__b__c__d_, 8);
+        let a__b__ca_db = op_or(a__b__c__d_, ______a__b_);
+        let ___a__b__ca = _mm_srli_si128(a__b__ca_db, 4);
+        let _______cadb = op_or(a__b__ca_db, ___a__b__ca);
+        _mm_cvtsi128_si32(_______cadb) as u32
+    }
+
+    #[target_feature(enable="sse3")]
+    unsafe fn compute_delta(curr: DataType, prev: DataType) -> DataType {
+        _mm_sub_epi32(curr,
+                      op_or(_mm_slli_si128(curr, 4),
+                            _mm_srli_si128(prev, 12))
+        )
+    }
+
+
+    #[target_feature(enable="sse3")]
+    #[allow(non_snake_case)]
+    unsafe fn integrate_delta(prev: DataType, delta: DataType) -> DataType {
+        let offset = _mm_shuffle_epi32(prev, 0xff);
+        let a__b__c__d_ = delta;
+        let ______a__b_ = _mm_slli_si128(delta, 8);
+        let a__b__ca_db = _mm_add_epi32(______a__b_, a__b__c__d_);
+        let ___a__b__ca = _mm_slli_si128(a__b__ca_db, 4);
+        let a_ab_abc_abcd: DataType = _mm_add_epi32(___a__b__ca, a__b__ca_db);
+        _mm_add_epi32(offset, a_ab_abc_abcd)
+    }
+
+    declare_bitpacker!(target_feature(enable="sse3"));
+
+    impl Available for UnsafeBitPackerImpl {
+        fn available() -> bool {
+            is_x86_feature_detected!("sse3")
+        }
+    }
+}
+
+mod scalar {
+
+    use super::BLOCK_LEN;
+    use Available;
+
+    type DataType = [u32; 4];
+
+
+    fn set1(el: i32) -> DataType {
+        [el as u32; 4]
+    }
+
+    fn right_shift_32(el: DataType, shift: i32) -> DataType {
+        [el[0] >> shift,
+         el[1] >> shift,
+         el[2] >> shift,
+         el[3] >> shift]
+    }
+
+    fn left_shift_32(el: DataType, shift: i32) -> DataType {
+        [el[0] << shift,
+         el[1] << shift,
+         el[2] << shift,
+         el[3] << shift]
+    }
+
+    fn op_or(left: DataType, right: DataType) -> DataType {
+        [left[0] | right[0],
+         left[1] | right[1],
+         left[2] | right[2],
+         left[3] | right[3]]
+    }
+
+    fn op_and(left: DataType, right: DataType) -> DataType {
+        [left[0] & right[0],
+         left[1] & right[1],
+         left[2] & right[2],
+         left[3] & right[3]]
+    }
+
+    unsafe fn load_unaligned(addr: *const DataType) -> DataType {
+        *addr
+    }
+
+    unsafe fn store_unaligned(addr: *mut DataType, data: DataType) {
+        *addr = data;
+    }
```

Comment:
The above two methods should either have their names change to use `aligned` instead of `unaligned`, or the implementations of these methods should use actual unaligned reads/writes. See https://doc.rust-lang.org/std/ptr/fn.read_unaligned.html and https://doc.rust-lang.org/std/ptr/fn.write_unaligned.html

Reply:
Nice catch. I randomly assumed pointer dereference was equivalent to `ptr::read_unaligned`.

Reply:
I need to fix some code in tantivy too (O_o)